sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -160,6 +170,40 @@ public boolean isSplittable() {
     }
   }
 
+  /** Describes a {@link DoFn.OnTimer} method. */
+  @AutoValue
+  public abstract static class OnTimerMethod implements DoFnMethod {
+    /** The annotated method itself. */
+    @Override
+    public abstract Method targetMethod();
+
+    /** The id on the method's {@link DoFn.TimerId} annotation. */
+    public abstract String id();
+
+    /** Types of optional parameters of the annotated method, in the order they appear. */
+    public abstract List<Parameter> extraParameters();
+
+    static OnTimerMethod create(Method targetMethod, String id, List<Parameter> extraParameters) {
+      return new AutoValue_DoFnSignature_OnTimerMethod(
+          targetMethod, id, Collections.unmodifiableList(extraParameters));
+    }
+  }
+
+  /**
+   * Describes a timer declaration; a field of type {@link TimerSpec} annotated with
+   * {@DoFn.TimerId}.
+   */
+  @AutoValue
+  public abstract static class TimerDeclaration {
+    public abstract String id();
+    public abstract Field field();
+
+    static TimerDeclaration create(String id, Field field) {
```

Comment:
I wonder if we should add a `DoFnField` class to parallel `DoFnMethod`. Also, for parallelism the `field` should be the first argument?